src/neo/SmartContract/InteropService.Contract.cs

```diff
@@ -56,50 +55,75 @@ private static bool Contract_Update(ApplicationEngine engine)
-                            engine.Snapshot.Storages.Add(new StorageKey
-                            {
-                                ScriptHash = hash_new,
-                                Key = key.Key
-                            }, new StorageItem
-                            {
-                                Value = value.Value,
-                                IsConstant = false
-                            });
-                        }
+                        oldcontract = engine.Snapshot.Contracts.GetAndChange(oldcontract.ScriptHash);
+                        oldcontract.IsUpdated = true;
+                    }
+                    else
+                    {
+                        Contract_UnAppend_Destroy(engine);
                     }
-                    Contract_Destroy(engine);
                 }
                 if (manifest.Length > 0)
                 {
-                    contract = engine.Snapshot.Contracts.GetAndChange(contract.ScriptHash);
-                    contract.Manifest = ContractManifest.Parse(manifest);
-                    if (!contract.Manifest.IsValid(contract.ScriptHash)) return false;
-                    if (!contract.HasStorage && engine.Snapshot.Storages.Find(engine.CurrentScriptHash.ToArray()).Any()) return false;
+                    newcontract = engine.Snapshot.Contracts.GetAndChange(newcontract.ScriptHash);
+                    newcontract.Manifest = ContractManifest.Parse(manifest);
+                    if (!newcontract.Manifest.IsValid(newcontract.ScriptHash)) return false;
+                    if (!newcontract.HasStorage && engine.Snapshot.Storages.Find(engine.CurrentScriptHash.ToArray()).Any()) return false;
                 }
 
                 return true;
             }
 
             private static bool Contract_Destroy(ApplicationEngine engine)
+            {
+                UInt160 hash = engine.CurrentScriptHash;
+                ContractState contract = engine.Snapshot.Contracts.TryGet(hash);
+                if (contract == null) return true;
+                if (!contract.RedirectionHash.Equals(UInt160.Zero))
+                {
+                    ContractState initcontract = engine.Snapshot.Contracts.TryGet(contract.RedirectionHash);
+                    if (initcontract != null)
+                    {
+                        engine.Snapshot.Contracts.Delete(contract.RedirectionHash);
+                        if (initcontract.HasStorage)
+                            foreach (var (key, _) in engine.Snapshot.Storages.Find(contract.RedirectionHash.ToArray()))
+                                engine.Snapshot.Storages.Delete(key);
+                    }
+                }
+                engine.Snapshot.Contracts.Delete(hash);
+                if (contract.HasStorage)
+                    foreach (var (key, _) in engine.Snapshot.Storages.Find(hash.ToArray()))
+                        engine.Snapshot.Storages.Delete(key);
+                return true;
+            }
+
```

Comment:
There are 3 places using the same code:
```csharp
 engine.Snapshot.Contracts.Delete(hash);
 if (contract.HasStorage)
         foreach (var (key, _) in engine.Snapshot.Storages.Find(hash.ToArray()))
                 engine.Snapshot.Storages.Delete(key);
```

I think we can abstract a delete method